### PR TITLE
fix(moe): allocate trtllm-gen expert_weights as bf16 in all launchers

### PR DESCRIPTION
**Pull request by aleozlx**
## 📌 Description

Follow-up to #3595 / #3644 (which fixed the FP4 op). The trtllm-gen routing kernel **always** writes expert weights in bfloat16 — `routingData.mDtypeOutput` is hard-set to `Bfloat16` in every branch of `Routing::Runner::run` (`csrc/trtllm_fused_moe_runner.cu`), independent of the `routing_logits` dtype.

But four launchers in `csrc/trtllm_fused_moe_kernel_launcher.cu` allocated the `expert_weights` output buffer from the **routing-logits** dtype (`dl_float32` when logits are fp32):

- `Bf16MoeLauncher`
- `Fp8PerTensorLauncher`
- `Fp8BlockScaleLauncher`
- `MxInt4BlockScaleLauncher`

The base `run()` returns that buffer verbatim on `do_finalize=false`, and the Python wrappers hand it straight back via `torch.from_dlpack`. So with fp32 routing logits (the conventional DeepSeekV3 case) + `do_finalize=false`, the caller receives **bf16 data mislabeled as fp32** and reads garbage — the same corruption #3644 fixed for FP4.

**Fix:** allocate `expert_weights` as `bfloat16` unconditionally to match the kernel's actual output. No-op for the common bf16-logits path and for `do_finalize=true` (internal workspace read back as bf16, just avoids a 2× over-allocation). Also removes the now-dead `mDtypeScore` member (write-only after this change).

**Why the fp32 branch existed:** a latent mismatch since #2534, which added a configurable score dtype (`mDtypeScore`) to support fp32 routing *logits*. But the expert-weights *output* (`mDtypeExpW`, later renamed `mDtypeOutput`) was always bf16 in that same PR — the launcher mistakenly sized/labeled the *output* buffer from `mDtypeScore` (an input/score knob). **fp32 logits remain fully supported** — they still flow to the kernel via `mRoutingLogitsDtype`/`mDtypeInput`; only the output-buffer dtype is corrected.

## 🔍 Related Issues

- Follow-up to #3595 / #3644 (same bug class, FP4 op)
- Latent mismatch originated in #2534

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Validation via GitHub CI (SM100-class trtllm-gen kernels). Suggested follow-up: a regression test that runs each of the four ops with fp32 DeepSeekV3 routing logits + `do_finalize=False` and asserts the returned `expert_weights.dtype == torch.bfloat16` (mirrors the FP4 test added in #3644).

## Reviewer Notes

Focus areas: confirm the Llama4 `token_scales` aliasing of this buffer is also fine as bf16, and that no caller depended on the (incorrect) fp32 labeling.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed routing output and expert weight buffers to consistently use the correct bfloat16 format.
  * Prevented potential data mislabeling when routing results are returned before finalization.
  * Improved consistency across supported Mixture-of-Experts execution modes.

### CODE REVIEW

File: csrc/trtllm_fused_moe_kernel_launcher.cu [hunk 1]
```diff
@@ -784,9 +774,12 @@ class Bf16MoeLauncher : public FusedMoeLauncher {
...
       workspace.expert_weights = const_cast<void*>(expert_weights.data_ptr());
     } else {
-      auto ew_dtype = mDtypeScore == btg::Dtype::Fp32 ? dl_float32 : dl_bfloat16;
+      // Allocate the routing-output buffer as bf16 to match the kernel's output
+      // (mDtypeOutput is always Bfloat16 in trtllm_fused_moe_runner.cu, never the
+      // logits dtype); a fp32 alloc would mislabel bf16 data when this buffer is
+      // surfaced to the caller verbatim on do_finalize=false. See #3595.
       FusedMoeLauncher::expert_weights =
-          alloc_tensor({args->num_tokens, args->top_k}, ew_dtype, hidden_states.device());
+          alloc_tensor({args->num_tokens, args->top_k}, dl_bfloat16, hidden_states.device());
```
[the remark this answers is not in the document]
Comment (aleozlx) [reply csrc/trtllm_fused_moe_kernel_launcher.cu:782]: Agreed on the coverage. Rather than a standalone test, we'll add this to the routing-aware unified MoE fuzzer, whose `routing_method × routing_input_mode × logits_dtype` axes are a natural fit for asserting the `do_finalize=False` expert_weights dtype across all four operators. Tracked in #3926.

**Comment by feih-nv** [comment]
LGTM. 
- Unconditional BF16 allocation matches every routing branch’s `mDtypeOutput`.
- Removing `mDtypeScore` is safe; FP32 logits still use `mRoutingLogitsDtype`.
- Llama4’s `token_scales` alias now points to the correctly typed BF16 routing output.
- The regression test covers only the FP4 op, while #3926 tracks the missing test coverage, which should include:
  - FP32 routing logits
  - do_finalize=False
  - assertion that expert_weights.dtype == torch.bfloat16
- Also, the unified-runner side of this same bug family is fixed in #3892.

**Review by nv-yunzheq — APPROVED** [review]
